File: day19/src/main.py
from pathlib import Path


def count_segmentations(string, segments, memo={'': 1}):
    if string in memo:
        return memo[string]

    count = memo[string] = sum(
        count_segmentations(string[len(segment):], segments, memo)
        for segment in segments
        if string.startswith(segment)
    )

    return count


def part1():
    data = Path('my_input.txt').read_text().strip()
    patterns, designs = data.split('\n\n')

    patterns = patterns.split(', ')
    designs = designs.splitlines()

    possible_count = sum(count_segmentations(design, patterns) > 0 for design in designs)

    print(f'Answer: {possible_count}')


def part2():
    data = Path('my_input.txt').read_text().strip()
    patterns, designs = data.split('\n\n')

    patterns = patterns.split(', ')
    designs = designs.splitlines()

    total_ways = sum(count_segmentations(design, patterns) for design in designs)

    print(f'Answer: {total_ways}')


if __name__ == '__main__':
    part1()
    part2()

File: day19/src/main.rs
use std::collections::HashMap;
use std::fs;

fn count_segmentations(string: &str, segments: &[&str]) -> usize {
    let mut memo: HashMap<String, usize> = HashMap::from([(String::new(), 1)]);

    fn inner(string: &str, segments: &[&str], memo: &mut HashMap<String, usize>) -> usize {
        if let Some(&count) = memo.get(string) {
            return count;
        }

        let count = segments
            .iter()
            .filter(|&&segment| string.starts_with(segment))
            .map(|&segment| inner(&string[segment.len()..], segments, memo))
            .sum();

        memo.insert(string.to_string(), count);
        count
    }

    inner(string, segments, &mut memo)
}

fn part1() {
    let input_data = fs::read_to_string("my_input.txt").expect("Failed to read input file");
    let parts: Vec<&str> = input_data.trim().split("\n\n").collect();
    let patterns: Vec<&str> = parts[0].split(", ").collect();
    let designs: Vec<&str> = parts[1].split('\n').collect();

    let possible_count = designs
        .iter()
        .filter(|design| count_segmentations(&design, &patterns) > 0)
        .count();

    println!("Answer: {}", possible_count);
}

fn part2() {
    let input_data = fs::read_to_string("my_input.txt").expect("Failed to read input file");
    let parts: Vec<&str> = input_data.trim().split("\n\n").collect();
    let patterns: Vec<&str> = parts[0].split(", ").collect();
    let designs: Vec<&str> = parts[1].split('\n').collect();

    let total_ways: usize = designs
        .iter()
        .map(|design| count_segmentations(&design, &patterns))
        .sum();

    println!("Answer: {}", total_ways);
}

fn main() {
    part1();
    part2();
}
